Update documentation (#17)
microsoft/proxy_wifi: README.md

    wifiProxy->Start();
```

## Architecture overview

Proxy_wifi is constituted of a service component (proxy_wifi service) meant to
run on a Windows host, and a driver component (proxy_wifi driver) meant to run
on a guest virtual machine. This repository contains the implementation of
proxy_wifi service for Windows. Currently, an implementation of proxy_wifi

> driver exists only for Linux

Reviewer: ```suggestion
driver exists only for Linux.
```

![General architecture](./doc/proxy_wifi_general_architecture.png)

Proxy_wifi driver role is to intercept Wi-Fi control path related operations
from the guest and forward them to proxy_wifi service. Communication between

> proxy_wifi driver and proxy_wifi service relies on two hv_sockets: one allows

Reviewer: Might want to say the unabbreviated term 'Hyper-V sockets (hv_sockets)' here since some people won't know the abbreviation. All following references can then just use the `hv_socket` abbreviation:

```suggestion
proxy_wifi driver and proxy_wifi service relies on two Hyper-V sockets (hvsockets): one allows
```

proxy_wifi driver to make a request to proxy_wifi service and wait for the
answer ; the other one allows proxy_wifi service to send spontaneous
notifications to proxy_wifi driver.

Proxy_wifi service is a library, its code will be running inside your
application process.  It uses the Windows Wlan API to perform any Wi-Fi

> realated operation in the host. It can also emulate operations on fake

Reviewer: ```suggestion
related operation in the host. It can also emulate operations on fake
```

networks if requested by the client application.

Proxy_wifi service present an observer interface (`ProxyWifiObserver`) that
client application should implement to be notified when an operation is request
by the guest or a network change is detected on the host. It is expected the
client application will monitor these events to synchronize proxy_wifi with
other component virtualizing the data path or L3 layer properties.

## Proxy_wifi service architecture

![Proxy_wifi service architecture](./doc/proxy_wifi_service_architecture.png)

Proxy_wifi is composed of two main layers: the transport
([Transport.hpp](lib/Transport.hpp)) and the operation handler
([OperationHandler.hpp](lib/OperationHandler.hpp)).

The concrete implementations of `ProxyWifiService` initialize these two layers
and expose a clean API to client applications.

> The transport is responsible of communications with the proxy_wifi driver on

Reviewer: ```suggestion
The transport is responsible for communications with the proxy_wifi driver on
```

the guest VM. The operation handler is responsible of honoring these requests
and notifying the client application.

### Transport layer

The transport layer communicate with the guest proxy_wifi driver over two channels:

- a request channel: it handles commands from the proxy_wifi driver and their
  responses, such as connection or scan requests.
- a notification channel: it handles spontaneous notification from the host, such
  as network disconnection, signal quality changes and scanned bss updates.

The transport layers deals with `Message` ([Messages.hpp](lib/Messages.hpp)) and isn't aware of their content or structure beside the header + body layout.

Messages coming over the request channel are handled as synchronous transactions: once a request is accepted in `Transport::AcceptConnections` ([Transport.hpp](lib/Transport.hpp)), no other request will be handled until the transaction is completed by sending an answer.

On the other hand, messages sent over the notification channel are handled in a fire and forget fashion: the transport send notification asynchronously through a queue and doesn't expect any answer from proxy_wifi driver.

> Two implementations of the transport layers are available: over HV sockets or over TCP sockets.

Reviewer: Is TCP actually still available? I thought we removed that.

Author: Since it's only a small amount of code when creating the socket, so we kept it as a potential debugging / test tool.

and to the client application.

The operation handler uses typed requests and responses, parsed from the
`Messages` used by the transport ([`Messages.hpp`](lib/Messages.hpp)).
Interface management, requests processing and notifications targetting the
guest are serialized through a work queue (`m_serializedRunner`).

Another work queue (`m_clientNotificationQueue`) serialize notifications sent
to the client application.

The operation handler maintain a list of `WlanInterface`
([WlanInterface.hpp](lib/WlanInterface.hpp)) that represent actual Wi-Fi
interfaces on the host or fake interfaces to simulate networks.

Requests are dispatched to interfaces, either serialized or running
concurrently depending on the operation type.

In particular:

- "Scan" operations are ran on all available interfaces concurrently, and the scan results are aggregated
- "Connect" operations are ran successively on each interface until one succeeds
- "Disconnect" operations target only the previously connected interface

Each interface handles the command depending on its concrete implementation,
using simulated data or fowarding the operation through the host Wlan API.

#### Simulated interface

> If the client application provide a `FakeNetworkProvider` callback when instancitating proxy_wifi service, a simulated (or "fake" interface will be created). It allows to simulate Wi-Fi networks in the guest.

Reviewer: ```suggestion
If the client application provide a `FakeNetworkProvider` callback when instantiating proxy_wifi service, a simulated (or "fake" interface will be created). It allows to simulate Wi-Fi networks in the guest.
```
